components/data-table/private/head.jsx

```diff
@@ -13,295 +13,304 @@ import HeaderCell from './header-cell';
+	indeterminateSelected: PropTypes.bool,
+	canSelectRows: PropTypes.oneOfType([
+		PropTypes.bool,
+		PropTypes.oneOf(['checkbox', 'radio']),
+	]),
+	columns: PropTypes.arrayOf(
+		PropTypes.shape({
+			Cell: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
+			props: PropTypes.object,
+		})
+	),
+	fixedHeader: PropTypes.bool,
+	id: PropTypes.string,
+	onToggleAll: PropTypes.func,
+	onSort: PropTypes.func,
+	showRowActions: PropTypes.bool,
+};
+
+const ActionsHeader = (props) => {
+	const tableContext = useContext(TableContext);
+	const cellContext = { columnIndex: props.columnIndex, rowIndex: 0 };
+	const { tabIndex, hasFocus, handleFocus, handleKeyDown } = useContextHelper(
+		tableContext,
+		cellContext,
+		props.fixedLayout
+	);
+	const { fixedHeader } = props;
+	const getContent = (style) => (
+		<div className="slds-th__action" style={style}>
+			<span className="slds-assistive-text">{props.assistiveText}</span>
+		</div>
+	);
 
-	// ### Prop Types
-	static propTypes = {
-		assistiveText: PropTypes.shape({
-			actionsHeader: PropTypes.string,
-			columnSort: PropTypes.string,
-			columnSortedAscending: PropTypes.string,
-			columnSortedDescending: PropTypes.string,
-			selectAllRows: PropTypes.string,
-			selectRow: PropTypes.string,
-		}),
-		allSelected: PropTypes.bool,
-		headerRefs: PropTypes.func,
-		isHidden: PropTypes.bool,
-		indeterminateSelected: PropTypes.bool,
-		canSelectRows: PropTypes.oneOfType([
-			PropTypes.bool,
-			PropTypes.oneOf(['checkbox', 'radio']),
-		]),
-		columns: PropTypes.arrayOf(
-			PropTypes.shape({
-				Cell: PropTypes.func,
-				props: PropTypes.object,
-			})
-		),
-		fixedHeader: PropTypes.bool,
-		id: PropTypes.string,
-		onToggleAll: PropTypes.func,
-		onSort: PropTypes.func,
-		showRowActions: PropTypes.bool,
+	return (
+		<th
+			className={classNames({ 'slds-has-focus': hasFocus })}
+			ref={(ref) => {
+				if (props.headerRefs) {
+					props.headerRefs(ref, 'action');
+				}
+				if (ref && hasFocus) {
+					ref.focus();
+				}
+			}}
+			scope="col"
+			style={{
+				height: fixedHeader ? 0 : null,
+				lineHeight: fixedHeader ? 0 : null,
+				width: '3.25rem',
+			}}
+			onFocus={handleFocus}
+			onKeyDown={handleKeyDown}
+			tabIndex={tabIndex}
+		>
+			{getContent(
+				fixedHeader
+					? {
+							height: 0,
+							overflow: 'hidden',
+							paddingBottom: 0,
+							paddingTop: 0,
+							visibility: 'hidden',
+					  }
+					: null
+			)}
+			{fixedHeader ? (
+				<CellFixed>
+					{getContent({
+						lineHeight: 1,
+						width: '100%',
+					})}
+				</CellFixed>
+			) : null}
+		</th>
+	);
+};
+
+const SelectHeader = (props) => {
+	const tableContext = useContext(TableContext);
+	const cellContext = { columnIndex: 0, rowIndex: 0 };
+	const { tabIndex, hasFocus, handleFocus, handleKeyDown } = useContextHelper(
+		tableContext,
+		cellContext,
+		props.fixedLayout
+	);
+	const { fixedHeader, canSelectRows } = props;
+	const getContent = (idSuffix, style, ariaHidden) => {
```

Comment:
does it make any perf difference to make such functions as callback or useMemo?  I noticed there are couple similar functions in other files as well

Reply:
this particular one is in the table head so it only renders O(1) per table as opposed to O(rowcount) so I left it as is